docs/source/key_components/rl_toolkit.rst

```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
+- ``DiscreteMADDPG``: MADDPG algorithm for policies that generate discrete actions.
 
-  shared_net_conf = {...}
-  actor_net_conf = {...}
-  critic_net_conf = {...}
-  shared_optim_conf = {torch.optim.SGD, {"lr": 0.0001}}
-  actor_optim_conf = (torch.optim.Adam, {"lr": 0.001})
-  critic_optim_conf = (torch.optim.RMSprop, {"lr": 0.001})
-
-  class MyACNet(DiscreteACNet):
-      def __init__(self):
-          super().__init__()
-          self.shared = FullyConnected(**shared_net_conf)
-          self.actor = FullyConnected(**actor_net_conf)
-          self.critic = FullyConnected(**critic_net_conf)
-          self.shared_optim = shared_optim_conf[0](self.shared.parameters(), **shared_optim_conf[1])
-          self.actor_optim = actor_optim_conf[0](self.actor.parameters(), **actor_optim_conf[1])
-          self.critic_optim = critic_optim_conf[0](self.critic.parameters(), **critic_optim_conf[1])
-
-      def forward(self, states, actor: bool = True, critic: bool = True):
-          representation = self.shared(states)
-          return (self.actor(representation) if actor else None), (self.critic(representation) if critic else None)
-
-      def step(self, loss):
-          self.shared_optim.zero_grad()
-          self.actor_optim.zero_grad()
-          self.critic_optim.zero_grad()
-          loss.backward()
-          self.shared_optim.step()
-          self.actor_optim.step()
-          self.critic_optim.step()
-
-To generate stochastic actions given a batch of states, call ``get_action`` on the model instance: 
+Each trainer has a corresponding ``Param`` class that used to manage all related parameters. For example,
+``DiscreteActorCriticParams`` contains all parameters used in ``DiscreteActorCritic``:
 
 .. code-block:: python
 
-  action, log_p, values = ac_model.get_action(state)
+   @dataclass
+   class DiscreteActorCriticParams(TrainerParams):
+       get_v_critic_net_func: Callable[[], VNet] = None
+       reward_discount: float = 0.9
+       grad_iters: int = 1
+       critic_loss_cls: Callable = None
+       clip_ratio: float = None
+       lam: float = 0.9
+       min_logp: Optional[float] = None
 
-To performing a single gradient step on the model, pass the loss to the ``step`` function: 
+An example of creating an actor-critic trainer:
 
 .. code-block:: python
 
-  ac_model.step(critic_loss + actor_loss)
+   DiscreteActorCritic(
+       name='ac',
+       params=DiscreteActorCriticParams(
+           device="cpu",
+           get_v_critic_net_func=lambda: MyCriticNet(state_dim=128),
+           reward_discount=.0,
+           grad_iters=10,
+           critic_loss_cls=torch.nn.SmoothL1Loss,
+           min_logp=None,
+           lam=.0
+       )
+   )
+
+In order to indicate which trainer each policy is trained by, in MARO, we require that the name of the policy
+starts with the name of the trainer responsible for training it. For example, policy ``ac_1.policy_1`` is trained
```

Comment:
starts -> start
https://forum.wordreference.com/threads/requires-that-something-be-read-grammar.3369520/